src/peer-routing.js

```diff
@@ -1,40 +1,126 @@
 'use strict'
 
 const errCode = require('err-code')
+const debug = require('debug')
+const log = debug('libp2p:peer-routing')
+log.error = debug('libp2p:peer-routing:error')
+
+const all = require('it-all')
 const pAny = require('p-any')
+const {
+  setDelayedInterval,
+  clearDelayedInterval
+} = require('set-delayed-interval')
```

Comment:
I created this module for us to use here, but also in:

- https://github.com/libp2p/js-libp2p/blob/0.30.x/src/circuit/index.js#L43-L88
- https://github.com/libp2p/js-libp2p-kad-dht/blob/master/src/random-walk.js#L46-L89

It is always tricky to implement a "setInterval" when the recurrent task is a promise. Other than that, if we need a delay beforehand, things get even more complex as we need to guarantee that the timer can be stopped for any of the pauses.

When looking at the `relay` logic I found a bug. If we clear the timeout when ` await this._libp2p.contentRouting.provide(cid)` is running, then the timeout will be assigned again afterward. We should be checking if we had timeout before creating the new one. This module aims to avoid introducing these issues when dealing with this use cases.

It would be great to have your 👀  on the module: https://github.com/vasco-santos/set-delayed-interval/pull/1

Reply:
Yeah, I'd really like to get rid of this whole boot delay logic. In reality we should probably have some async boot function and when that finishes these other systems initiate, but I think we can revisit this when we update the config logic. I'll review the module

Reply:
It would be great to orchestrate subsystems like that. I will add this comment to the configuration issue for us to remember